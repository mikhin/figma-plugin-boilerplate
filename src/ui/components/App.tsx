import { useStore } from '@nanostores/react';
import { MessageTypes } from '~types/messages.ts';
import {
  $isUserSelectionEmpty,
  $userSelection,
} from '~ui/stores/user-selection.ts';
import React, { useEffect } from 'react';

export const App: React.FC = () => {
  const userSelection = useStore($userSelection);
  const isUserSelectionEmpty = useStore($isUserSelectionEmpty);

  useEffect(() => {
    parent.postMessage(
      {
        pluginMessage: {
          type: MessageTypes.UiReady,
        },
      },
      '*'
    );
  }, []);

  return (
    <div className="p-4">
      {isUserSelectionEmpty ? (
        <div>
          <h1 className="pb-2 text-xl font-bold text-gray-900">
            Select some nodes
          </h1>
        </div>
      ) : (
        <>
          <h1 className="mb-2 border-b border-gray-200 pb-2 text-xl font-bold text-gray-900">
            Selected nodes names
          </h1>

          <ul>
            {userSelection.selectedNodeNames.map((name) => (
              <li
                className="border-b border-gray-200 py-2 text-sm font-medium text-gray-900 last:border-b-0"
                key={name}
              >
                {name}
              </li>
            ))}
          </ul>
        </>
      )}
    </div>
  );
};
