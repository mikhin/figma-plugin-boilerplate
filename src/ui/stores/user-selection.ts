import {
  type Message,
  MessageTypes,
  type SelectionChangePayload,
} from '~types/messages.ts';
import { atom, computed, onMount } from 'nanostores';

export const $userSelection = atom<SelectionChangePayload>({
  selectedNodeNames: [],
});

export const $isUserSelectionEmpty = computed($userSelection, (selection) => {
  return selection.selectedNodeNames.length === 0;
});

onMount($userSelection, () => {
  const addMessageListener = (
    event: MessageEvent<Message<SelectionChangePayload>>
  ): void => {
    if (event.data?.pluginMessage.type === MessageTypes.SelectionChange) {
      $userSelection.set(event.data.pluginMessage.payload);
    }
  };

  window.addEventListener('message', addMessageListener);

  return () => {
    window.removeEventListener('message', addMessageListener);
  };
});
