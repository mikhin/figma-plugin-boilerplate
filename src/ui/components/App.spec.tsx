import { render } from '@testing-library/react';
import { App } from '~ui/components/App.tsx';
import { $userSelection } from '~ui/stores/user-selection.ts';
import { describe, expect, test } from 'vitest';

describe('Layout component', () => {
  test('Render empty list', () => {
    expect(render(<App />).container).toMatchSnapshot();
  });

  test('Render list with content', () => {
    $userSelection.set({
      selectedNodeNames: ['Frame 1', 'Frame 2'],
    });

    expect(render(<App />).container).toMatchSnapshot();
  });
});
