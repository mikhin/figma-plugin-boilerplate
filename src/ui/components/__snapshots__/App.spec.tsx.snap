// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Layout component > Render empty list 1`] = `
<div>
  <div
    class="p-4"
  >
    <div>
      <h1
        class="pb-2 text-xl font-bold text-gray-900"
      >
        Select some nodes
      </h1>
    </div>
  </div>
</div>
`;

exports[`Layout component > Render list with content 1`] = `
<div>
  <div
    class="p-4"
  >
    <h1
      class="mb-2 border-b border-gray-200 pb-2 text-xl font-bold text-gray-900"
    >
      Selected nodes names
    </h1>
    <ul>
      <li
        class="border-b border-gray-200 py-2 text-sm font-medium text-gray-900 last:border-b-0"
      >
        Frame 1
      </li>
      <li
        class="border-b border-gray-200 py-2 text-sm font-medium text-gray-900 last:border-b-0"
      >
        Frame 2
      </li>
    </ul>
  </div>
</div>
`;
